File: app/src/main/java/com/nfcalarmclock/activealarm/NacActiveAlarmNotification.java
package com.nfcalarmclock.activealarm;

import android.annotation.TargetApi;
import android.app.NotificationChannel;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Build;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;

import com.nfcalarmclock.alarm.NacAlarm;
import com.nfcalarmclock.system.NacCalendar;
import com.nfcalarmclock.system.NacIntent;
import com.nfcalarmclock.util.notification.NacNotification;
import com.nfcalarmclock.shared.NacSharedConstants;
import com.nfcalarmclock.R;
import com.nfcalarmclock.nfc.NacNfc;

import java.lang.Float;
import java.util.Calendar;
import java.util.Locale;

/**
 * Notification to display for active alarms.
 */
public class NacActiveAlarmNotification
	extends NacNotification
{

	/**
	 * Group key.
	 */
	public static final String GROUP = "NacNotiGroupActiveAlarm";

	/**
	 * Alarm.
	 */
	private NacAlarm mAlarm;

	/**
	 */
	public NacActiveAlarmNotification(Context context)
	{
		super(context);

		this.mAlarm = null;
	}

	/**
	 * @see NacNotification#builder()
	 */
	@Override
	protected NotificationCompat.Builder builder()
	{
		Bitmap icon = this.getIcon();
		PendingIntent activityPending = this.getContentPendingIntent();
		PendingIntent snoozePending = this.getSnoozePendingIntent();
		PendingIntent dismissPending = this.getDismissPendingIntent(activityPending);
		NacSharedConstants cons = this.getSharedConstants();

		return super.builder()
			.setLargeIcon(icon)
			.setFullScreenIntent(activityPending, true)
			.setAutoCancel(false)
			.setOngoing(true)
			.setShowWhen(true)
			.setTicker(cons.getAppName())
			.addAction(R.mipmap.snooze, cons.getActionSnooze(), snoozePending)
			.addAction(R.mipmap.dismiss, cons.getActionDismiss(), dismissPending);
	}

	/**
	 */
	@TargetApi(Build.VERSION_CODES.O)
	@Override
	protected NotificationChannel createChannel()
	{
		NotificationChannel channel = super.createChannel();
		if (channel == null)
		{
			return null;
		}

		channel.setShowBadge(true);
		channel.enableLights(true);
		channel.enableVibration(true);
		return channel;
	}

	/**
	 * @return The alarm.
	 */
	public NacAlarm getAlarm()
	{
		return this.mAlarm;
	}

	/**
	 * @return The notification description.
	 */
	protected String getChannelDescription()
	{
		NacSharedConstants cons = this.getSharedConstants();
		return cons.getDescriptionActiveNotification();
	}

	/**
	 * @return The notification name.
	 */
	protected String getChannelName()
	{
		NacSharedConstants cons = this.getSharedConstants();
		return cons.getActiveNotification();
	}

	/**
	 * @see NacNotification#getChannelId()
	 */
	protected String getChannelId()
	{
		return "NacNotiChannelActiveAlarm";
	}

	/**
	 * @see NacNotification#getContentPendingIntent()
	 */
	protected PendingIntent getContentPendingIntent()
	{
		Context context = this.getContext();
		NacAlarm alarm = this.getAlarm();
		long id = (alarm != null) ? alarm.getId() : 0;
		Intent intent = NacIntent.createAlarmActivity(context, alarm);

		return PendingIntent.getActivity(context, (int)id, intent,
			PendingIntent.FLAG_UPDATE_CURRENT);
	}

	/**
	 * @see NacNotification#getContentText()
	 */
	public String getContentText()
	{
		Context context = this.getContext();
		NacAlarm alarm = this.getAlarm();
		Calendar now = Calendar.getInstance();
		String time = NacCalendar.Time.getFullTime(context, now);
		String name = alarm.getName();
		Locale locale = Locale.getDefault();

		return name.isEmpty() ? time
			: String.format(locale, "%1$s  —  %2$s", time, name);
	}

	/**
	 * @return The pending intent to use when dismissing the alarm.
	 */
	private PendingIntent getDismissPendingIntent(
		PendingIntent activityPendingIntent)
	{
		Context context = this.getContext();
		NacAlarm alarm = this.getAlarm();

		if (NacNfc.shouldUseNfc(context, alarm))
		{
			return activityPendingIntent;
		}

		Intent intent = NacIntent.dismissForegroundService(context, alarm);

		return PendingIntent.getService(context, 0, intent,
			PendingIntent.FLAG_CANCEL_CURRENT);
	}

	/**
	 * @see NacNotification#getGroup()
	 */
	protected String getGroup()
	{
		return GROUP;
	}

	/**
	 * @return The notification large icon.
	 */
	protected Bitmap getIcon()
	{
		NacSharedConstants cons = this.getSharedConstants();
		Resources res = cons.getResources();
		Bitmap icon = BitmapFactory.decodeResource(res, R.mipmap.app);
		float density = res.getDisplayMetrics().density;
		float size;

		if (Float.compare(density, 4.0f) >= 0)
		{
			size = 256f;
		}
		else if (Float.compare(density, 3.0f) >= 0)
		{
			size = 192f;
		}
		else if (Float.compare(density, 2.0f) >= 0)
		{
			size = 128f;
		}
		else if (Float.compare(density, 1.5f) >= 0)
		{
			size = 96f;
		}
		else if (Float.compare(density, 1.0f) >= 0)
		{
			size = 64f;
		}
		else if (Float.compare(density, 0.75f) >= 0)
		{
			size = 48f;
		}
		else
		{
			return icon;
		}

		return Bitmap.createScaledBitmap(icon, (int)(size*density),
			(int)(size*density), true);
	}

	/**
	 * @see NacNotification#getId()
	 */
	protected int getId()
	{
		return 79;
	}

	/**
	 * @see NacNotification#getImportance()
	 */
	protected int getImportance()
	{
		return NotificationManagerCompat.IMPORTANCE_HIGH;
	}

	/**
	 * @see NacNotification#getPriority()
	 */
	protected int getPriority()
	{
		return NotificationCompat.PRIORITY_MAX;
	}

	/**
	 * @return The notification title.
	 */
	public String getTitle()
	{
		Locale locale = Locale.getDefault();
		NacSharedConstants cons = this.getSharedConstants();

		return String.format(locale, "<b>%s</b>", cons.getAppName());
	}

	/**
	 * @return The pending intent to use when snoozing.
	 */
	private PendingIntent getSnoozePendingIntent()
	{
		Context context = this.getContext();
		NacAlarm alarm = this.getAlarm();
		Intent intent = NacIntent.snoozeForegroundService(context, alarm);

		return PendingIntent.getService(context, 0, intent,
			PendingIntent.FLAG_CANCEL_CURRENT);
	}

	/**
	 * Set the alarm.
	 */
	public void setAlarm(NacAlarm alarm)
	{
		this.mAlarm = alarm;
	}

}

File: app/src/main/java/com/nfcalarmclock/activealarm/NacActiveAlarmNotification.kt
package com.nfcalarmclock.activealarm

import android.annotation.TargetApi
import android.app.NotificationChannel
import android.app.PendingIntent
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import com.nfcalarmclock.R
import com.nfcalarmclock.alarm.db.NacAlarm
import com.nfcalarmclock.nfc.NacNfc
import com.nfcalarmclock.util.NacCalendar
import com.nfcalarmclock.util.NacIntent.createAlarmActivity
import com.nfcalarmclock.util.NacIntent.dismissForegroundService
import com.nfcalarmclock.util.NacIntent.snoozeForegroundService
import com.nfcalarmclock.view.notification.NacNotification
import java.util.Calendar
import java.util.Locale

/**
 * Notification to display for active alarms.
 */
class NacActiveAlarmNotification(context: Context)
	: NacNotification(context)
{

	/**
	 * The alarm to show the notification about.
	 */
	var alarm: NacAlarm? = null

	/**
	 * @see NacNotification.id
	 */
	override val id: Int
		get() = Companion.ID

	/**
	 * @see NacNotification.channelId
	 */
	override val channelId: String
		get() = "NacNotiChannelActiveAlarm"

	/**
	 * @see NacNotification.channelName
	 */
	override val channelName: String
		get() = context.getString(R.string.title_active_alarm)

	/**
	 * @see NacNotification.channelDescription
	 */
	override val channelDescription: String
		get() = context.getString(R.string.description_active_alarm)

	/**
	 * @see NacNotification.title
	 */
	override val title: String
		get()
		{
			val locale = Locale.getDefault()

			return String.format(locale, "<b>%s</b>", appName)
		}

	/**
	 * @see NacNotification.priority
	 */
	override val priority: Int
		get() = NotificationCompat.PRIORITY_MAX

	/**
	 * @see NacNotification.importance
	 */
	override val importance: Int
		get() = NotificationManagerCompat.IMPORTANCE_HIGH

	/**
	 * @see NacNotification.group
	 */
	override val group: String
		get() = "NacNotiGroupActiveAlarm"

	/**
	 * @see NacNotification.contentText
	 */
	override val contentText: String
		get()
		{
			val now = Calendar.getInstance()
			val time = NacCalendar.Time.getFullTime(context, now)
			val locale = Locale.getDefault()
			val name = alarm?.name ?: ""

			return if (name.isEmpty()) time else String.format(locale,
				"%1\$s  —  %2\$s", time, name)
		}

	/**
	 * @see NacNotification.contentPendingIntent
	 */
	override val contentPendingIntent: PendingIntent
		get()
		{
			val id = alarm?.id ?: 0
			val intent = createAlarmActivity(context, alarm)

			// Determine the pending intent flags
			var flags = PendingIntent.FLAG_UPDATE_CURRENT

			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
			{
				flags = flags or PendingIntent.FLAG_IMMUTABLE
			}

			// Return the pending intent for the activity
			return PendingIntent.getActivity(context, id.toInt(), intent, flags)
		}

	/**
	 * The pending intent to use when snoozing.
	 */
	private val snoozePendingIntent: PendingIntent
		get()
		{
			val intent = snoozeForegroundService(context, alarm)

			// Determine the pending intent flags
			var flags = PendingIntent.FLAG_CANCEL_CURRENT

			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
			{
				flags = flags or PendingIntent.FLAG_IMMUTABLE
			}

			// Create the pending intent
			return PendingIntent.getService(context, 0, intent, flags)
		}

	/**
	 * @return The notification large icon.
	 */
	protected val largeIcon: Bitmap?
		get()
		{
			val res = context.resources
			val icon = BitmapFactory.decodeResource(res, R.mipmap.app)
			val density = res.displayMetrics.density

			// Determine the size of the bitmap
			val size: Float = if (density.compareTo(4.0f) >= 0)
			{
				256f
			}
			else if (density.compareTo(3.0f) >= 0)
			{
				192f
			}
			else if (density.compareTo(2.0f) >= 0)
			{
				128f
			}
			else if (density.compareTo(1.5f) >= 0)
			{
				96f
			}
			else if (density.compareTo(1.0f) >= 0)
			{
				64f
			}
			else if (density.compareTo(0.75f) >= 0)
			{
				48f
			}
			else
			{
				return icon
			}

			// Determine the integer size
			val actualSize = (size * density).toInt()

			// Create the bitmap
			return if (icon != null)
			{
				Bitmap.createScaledBitmap(icon, actualSize, actualSize, true)
			}
			else
			{
				null
			}
		}

	/**
	 * Name of the app.
	 */
	private val appName = context.getString(R.string.app_name)

	/**
	 * @see NacNotification.builder
	 */
	public override fun builder(): NotificationCompat.Builder
	{
		// Create the dismiss pending intent
		val dismissPending = getDismissPendingIntent(contentPendingIntent)

		// Notification actions
		val dismiss = context.getString(R.string.action_alarm_dismiss)
		val snooze = context.getString(R.string.action_alarm_snooze)

		// Build the notification
		return super.builder()
			.setLargeIcon(largeIcon)
			.setFullScreenIntent(contentPendingIntent, true)
			.setForegroundServiceBehavior(NotificationCompat.FOREGROUND_SERVICE_IMMEDIATE)
			.setAutoCancel(false)
			.setOngoing(true)
			.setShowWhen(true)
			.setTicker(appName)
			.addAction(R.mipmap.snooze, snooze, snoozePendingIntent)
			.addAction(R.mipmap.dismiss, dismiss, dismissPending)
	}

	/**
	 * @see NacNotification.createChannel
	 */
	@TargetApi(Build.VERSION_CODES.O)
	override fun createChannel(): NotificationChannel
	{
		// Create the channel
		val channel = super.createChannel()

		// Setup the channel
		channel.setShowBadge(true)
		channel.enableLights(true)
		channel.enableVibration(true)

		return channel
	}

	/**
	 * Get the pending intent to use when dismissing the alarm.
	 *
	 * @return The pending intent to use when dismissing the alarm.
	 */
	private fun getDismissPendingIntent(activityPendingIntent: PendingIntent)
		: PendingIntent
	{
		// NFC should be used so show the active alarm activity pending intent
		if (NacNfc.shouldUseNfc(context, alarm))
		{
			return activityPendingIntent
		}

		// Create an intent to dismiss the active alarm service
		val intent = dismissForegroundService(context, alarm)

		// Determine the pending intent flags
		var flags = PendingIntent.FLAG_CANCEL_CURRENT

		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
		{
			flags = flags or PendingIntent.FLAG_IMMUTABLE
		}

		// Create the pending intent
		return PendingIntent.getService(context, 0, intent, flags)
	}

	companion object
	{

		/**
		 * Notification ID.
		 */
		const val ID = 79

	}

}